packages/lodestar/src/api/impl/beacon/state/utils.ts

```diff
@@ -182,3 +185,46 @@ export function filterStateValidatorsByStatuses(
   }
   return responses;
 }
+
+/**
+ * Get the archived state nearest to `slot`.
+ */
+function getNearestArchivedState(
+  stateCache: StateContextCache,
+  forkChoice: IForkChoice,
+  slot: Slot
+): CachedBeaconState<allForks.BeaconState> {
+  // @TODO: use config instead of hardcoding 32
+  const nearestArchivedStateSlot = slot - (slot % (PERSIST_STATE_EVERY_EPOCHS * 32));
+
+  let state: CachedBeaconState<allForks.BeaconState> | null = null;
+  const blockSummary = forkChoice.getCanonicalBlockSummaryAtSlot(nearestArchivedStateSlot);
+  if (blockSummary) {
+    state = stateCache.get(blockSummary.stateRoot);
+  }
+  if (state === null) throw new Error("getNearestArchivedState: cannot find state");
+  return state;
+}
+
+async function getFinalizedState(
+  db: IBeaconDb,
+  stateCache: StateContextCache,
+  forkChoice: IForkChoice,
+  slot: Slot
+): Promise<CachedBeaconState<allForks.BeaconState>> {
+  // @TODO: we should probably use SLOTS_PER_EPOCH instead of hard-coding `32`, but i'm not sure how without passing the config all the way from all the upstream calls
```

Comment:
Yeah, you can just add `config: IBeaconConfig` as the first param in this function.

Perhaps we can make a public `IBeaconChain.config` property, so it can be retrieved from `chain` in `resolveStateId`